Use of deprecated PyUnicode_GetSize in Python module (fails build with Python 3.7)

Build of dionaea on Fedora Rawhide fails at present, with this error:

    gcc -pthread -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -Wno-error -D_GNU_SOURCE -std=c99 -fPIC -D_GNU_SOURCE -DPYEV_VERSION="0.9.0" -UNDEBUG -I../../include -I../../ -I/usr/include/glib-2.0 -I/usr/include/python3.7m -I/usr/lib64/glib-2.0/include -I/usr/include/python3.7m -c binding.c -o build/temp.linux-x86_64-3.7/binding.o -DEV_COMPAT3=0 -Wall -Werror -fPIC -fno-strict-aliasing -g -pthread
    gcc -pthread -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -Wno-error -D_GNU_SOURCE -std=c99 -fPIC -D_GNU_SOURCE -DPYEV_VERSION="0.9.0" -UNDEBUG -I../../include -I../../ -I/usr/include/glib-2.0 -I/usr/include/python3.7m -I/usr/lib64/glib-2.0/include -I/usr/include/python3.7m -c module.c -o build/temp.linux-x86_64-3.7/module.o -DEV_COMPAT3=0 -Wall -Werror -fPIC -fno-strict-aliasing -g -pthread
    module.c: In function ‘pyobjectstring’:
    module.c:988:2: error: ‘PyUnicode_GetSize’ is deprecated [-Werror=deprecated-declarations]
      Py_ssize_t pysize = PyUnicode_GetSize(pyobjectstr);
      ^~~~~~~~~~
    In file included from /usr/include/python3.7m/Python.h:80,
                     from module.c:30:
    /usr/include/python3.7m/unicodeobject.h:793:24: note: declared here
     PyAPI_FUNC(Py_ssize_t) PyUnicode_GetSize(
                            ^~~~~~~~~~~~~~~~~
    cc1: all warnings being treated as errors

It seems in Python 3.7, `PyUnicode_GetSize` is deprecated, and as the dionaea build process uses `-Werror`, this fails the build.

--- a/modules/python/module.c
+++ b/modules/python/module.c
@@ -972,18 +972,18 @@ static char *pyobjectstring(PyObject *obj)
 	if( PyType_Check(obj) )
 		return g_strdup(((PyTypeObject* ) obj)->tp_name);
 
-	if( PyUnicode_Check(obj) )
+	if( PyUnicode_Check(obj) ) {
 		pyobjectstr = obj;
-	else 
-	if( (pyobjectstr = PyObject_Repr(obj)) != NULL )
-	{
-		if( PyUnicode_Check(pyobjectstr) == 0 )
-		{
-			Py_XDECREF(pyobjectstr);
-			return g_strdup("<!utf8>");
+	} else {
+		if( (pyobjectstr = PyObject_Repr(obj)) != NULL ) {
+			if( PyUnicode_Check(pyobjectstr) == 0 ) {
+				Py_XDECREF(pyobjectstr);
+				return g_strdup("<!utf8>");
+			}
+		} else {
+			return g_strdup("<!repr>");
 		}
-	} else
-		return g_strdup("<!repr>");
+	}
 
 	Py_ssize_t pysize = PyUnicode_GetSize(pyobjectstr);
 	wchar_t * str = (wchar_t *) malloc((pysize + 1) * sizeof(wchar_t));

--- a/modules/python/module.c
+++ b/modules/python/module.c
@@ -985,10 +985,8 @@ static char *pyobjectstring(PyObject *obj)
 		}
 	}
 
-	Py_ssize_t pysize = PyUnicode_GetSize(pyobjectstr);
-	wchar_t * str = (wchar_t *) malloc((pysize + 1) * sizeof(wchar_t));
-	PyUnicode_AsWideChar(pyobjectstr, str, pysize);
-	str[pysize] = '\0';
+	Py_ssize_t pysize;
+	wchar_t *str = PyUnicode_AsWideCharString(pyobjectstr, &pysize);
 
 	if( pyobjectstr != obj )
 		Py_DECREF(pyobjectstr);
@@ -1002,7 +1000,7 @@ static char *pyobjectstring(PyObject *obj)
 
 	// convert
 	wcstombs(cstr, str, csize + 1);
-	free(str);
+	PyMem_Free(str);
 	return cstr;
 }
 

